server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsageTelemetry.java

```diff
@@ -8,172 +8,174 @@
 
-    private final LongAdder totalCCSCount;  // TODO: need this? or just sum the successfulSearchTelem and failedSearchTelem ?
-    private final SuccessfulCCSTelemetry successfulSearchTelem;
-    private final FailedCCSTelemetry failedSearchTelem;
+    /**
+     * Result of the request execution.
+     * Either "success" or a failure reason.
+     */
+    public enum Result {
+        SUCCESS("success"),
+        REMOTES_UNAVAILABLE("remotes_unavailable"),
+        CANCELED("canceled"),
+        // TODO: do we need this? If we subtract all known reasons, the rest is unknown.
+        // May be helpful though if there's a lot of other reasons and it may be hard to calculate the unknowns for some clients.
+        UNKNOWN("unknown");
+
+        private final String name;
+
+        Result(String name) {
+            this.name = name;
+        }
+
+        public String getName() {
+            return name;
+        }
+    }
+
+    // Not enum because we won't mind other places adding their own features
+    public static final String MRT_FEATURE = "mrt_on";
+    public static final String ASYNC_FEATURE = "async";
+    public static final String WILDCARD_FEATURE = "wildcards";
+    // TODO: Do we need to count unknown clients separately?
+    // Again, total - known = unknown
+    private static final String CLIENT_UNKNOWN = "unknown";
+
+    // TODO: do we need LongAdder here or long is enough? Since updateUsage is synchronized, worst that can happen is
+    // we may miss a count on read.
+    private final LongAdder totalCount;
+    private final LongAdder successCount;
+    private final Map<Result, LongAdder> failureReasons;
+
+    private final LongMetric took;
+    private final LongMetric tookMrtTrue;
+    private final LongMetric tookMrtFalse;
+    private final LongMetric remotesPerSearch;
+    private final LongAdder skippedRemotes;
+
+    private final Map<String, LongAdder> featureCounts;
+
+    private final Map<String, LongAdder> clientCounts;
     private final Map<String, PerClusterCCSTelemetry> byRemoteCluster;
 
     public CCSUsageTelemetry() {
-        this.totalCCSCount = new LongAdder();
-        this.successfulSearchTelem = new SuccessfulCCSTelemetry();
-        this.failedSearchTelem = new FailedCCSTelemetry();
         this.byRemoteCluster = new ConcurrentHashMap<>();
+        totalCount = new LongAdder();
+        successCount = new LongAdder();
+        failureReasons = new ConcurrentHashMap<>();
+        took = new LongMetric();
+        tookMrtTrue = new LongMetric();
+        tookMrtFalse = new LongMetric();
+        remotesPerSearch = new LongMetric();
+        skippedRemotes = new LongAdder();
+        featureCounts = new ConcurrentHashMap<>();
+        clientCounts = new ConcurrentHashMap<>();
     }
 
     public void updateUsage(CCSUsage ccsUsage) {
+        // Ignore empty usage objects, we only care for true cross-cluster requests
+        if (ccsUsage.getRemotesCount() == 0) {
```

Comment:
Should this ever happen? If no, then just add an assert here that remotes count > 0.

Reply:
It shouldn't. I think you're right, assert here is better, I'll change it.